contracts/types/YieldSplitter.sol

```diff
@@ -0,0 +1,167 @@
+    function index() external view returns (uint256 index);
+}
+
+/**
+    @title YieldSplitter
+    @notice Abstract contract that allows users to create deposits for their gOHM and have
+            their yield claimable by the specified recipient party. This contract's functions
+            are designed to be as generic as possible. This contract's responsibility is
+            the accounting of the yield splitting. All other logic such as error handling,
+            emergency controls, sending and recieving gOHM is up to the implementation of
+            this abstract contract to handle.
+ */
+abstract contract YieldSplitter {
+    using SafeERC20 for IERC20;
+
+    error YieldSplitter_NotYourDeposit();
+
+    IOHMIndexWrapper public immutable indexWrapper;
+
+    struct DepositInfo {
+        uint256 id;
+        address depositor;
+        uint256 principalAmount; // Total amount of sOhm deposited as principal, 9 decimals.
+        uint256 agnosticAmount; // Total amount deposited priced in gOhm. 18 decimals.
+    }
+
+    uint256 public idCount;
+    mapping(uint256 => DepositInfo) public depositInfo; // depositId -> DepositInfo
+    mapping(address => uint256[]) public depositorIds; // address -> Array of the deposit id's deposited by user
+
+    /**
+        @notice Constructor
+        @param indexWrapper_ Address of contract that will return the sOHM to gOHM index. 
+                             On mainnet this will be sOHM but on other chains can be an oracle wrapper.
+    */
+    constructor(address indexWrapper_) {
+        indexWrapper = IOHMIndexWrapper(indexWrapper_);
+    }
+
+    /**
+        @notice Create a deposit.
+        @param depositor_ Address of depositor
+        @param amount_ Amount in gOhm. 18 decimals.
+    */
+    function _deposit(address depositor_, uint256 amount_) internal returns (uint256 depositId) {
+        depositorIds[depositor_].push(idCount);
+
+        depositInfo[idCount] = DepositInfo({
+            id: idCount,
+            depositor: depositor_,
+            principalAmount: _fromAgnostic(amount_),
+            agnosticAmount: amount_
+        });
+
+        depositId = idCount;
+        idCount++;
+    }
+
+    /**
+        @notice Add more gOhm to the depositor's principal deposit.
+        @param id_ Id of the deposit.
+        @param amount_ Amount of gOhm to add. 18 decimals.
+    */
+    function _addToDeposit(uint256 id_, uint256 amount_) internal {
+        DepositInfo storage userDeposit = depositInfo[id_];
+        userDeposit.principalAmount += _fromAgnostic(amount_);
+        userDeposit.agnosticAmount += amount_;
+    }
+
+    /**
+        @notice Withdraw part of the principal amount deposited.
+        @param id_ Id of the deposit.
+        @param amount_ Amount of gOHM to withdraw.
+    */
+    function _withdrawPrincipal(uint256 id_, uint256 amount_) internal {
+        if (depositInfo[id_].depositor != msg.sender) revert YieldSplitter_NotYourDeposit();
```

Comment:
would this not read the YieldDirectorV2 address as msg.sender? or is it different when inheriting abstracts?